### PR TITLE
[CALCITE-4368] TopDownOptTest fails if applying non-substitution rule first

### PR DESCRIPTION
**Pull request by chunweilei**
Usually O_INPUTS are only applied for groups with physical convention. But
when enabling AbstractConverter, the input of AbstractConverter might be
a group with NONE convention. In that case, no need to apply O_INPUTS.
Otherwise, it might throw an exception due to impossible transformation(
physical convention -> none convention).

Other change:
* some cosmetic fix-ups
* print the upper bound of the RelSubSet

### CODE REVIEW

**Review by FatLittle — APPROVED** [review]

**Comment by danny0405** [comment]
Remember to update the commit message before commit.

**Comment by chunweilei** [comment]
> emember to update the commit message before commit.

Could you please tell me what I should update about the commit message? Thanks.

**Comment by chunweilei** [comment]
Updated the commit message. Thank you for your suggestion, @danny0405 .

**Comment by vlsi** [comment]
Frankly speaking, it is hard to review the change because there are lots of comment changes, and it is hard to spot code changes.

Please refrain from mixing unrelated comment changes and code changes into a single commit.

It does not mean I require that for this case, however, I opened this PR several times, and every time I thought it was javadoc-only change.

**Comment by vlsi** [comment]
> Could you please tell me what I should update about the commit message? Thanks

@chunweilei , could you please clarify what is the nature of the change?
Is it a cosmetic change?
Does this PR fix a bug?
Does this PR introduce a new feature?

The commit message says `[CALCITE-4368] TopDownOptTest fails if applying non-substitution rule first`, so the commit message suggests the change fixes a bug. However, the diff looks like a javadoc-only cosmetic change.

In case the PR is cosmetic, then the very same thing should be mentioned in the commit message. Otherwise, it is misleading, and it is time-consuming for the reviewers.

**Comment by chunweilei** [comment]
@vlsi I have updated the commit message. This change not only fixes a bug but also have some fix-ups.

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 1]
```diff
@@ -311,31 +312,39 @@ default boolean onProduce(RelNode node) {
...
         }
       }
-      // always apply O_INPUTS first so as to get an valid upper bound
+
+      // No need to apply O_INPUTS if the group has not been implemented yet.
+      // It only happens when enabling AbstractConverter.
+      if (group.getConvention() == Convention.NONE) {
+        LOGGER.debug("Skip optimizing because of none convention: {}", group);
+        return;
+      }
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:345]: @vlsi Here is the main change.

**Comment by chunweilei** [comment]
> Frankly speaking, it is hard to review the change because there are lots of comment changes, and it is hard to spot code changes.
> 
> Please refrain from mixing unrelated comment changes and code changes into a single commit.
> 
> It does not mean I require that for this case, however, I opened this PR several times, and every time I thought it was javadoc-only change.

I don't want to see many commits that only have comment changes. It is not necessary. I should have underlined the main change. Then it is much easier for reviewers.

**Comment by vlsi** [comment]
@chunweilei, please split the change in two:
* code change with the corresponding comments
* other comment cleanups

Currently, it is hard to figure out where is the code part of the change.

**Comment by chunweilei** [comment]
The main change is here: https://github.com/apache/calcite/pull/2237/files#diff-ad65441646b31d60e7a55d6bac2c1f4d9ddba3473b227963800b1b2bc292a319R342

**Comment by vlsi** [comment]
Please add tests to cover the change

**Comment by chunweilei** [comment]
> Please add tests to cover the change

It's difficult to add a test. But you can reproduce it if you change the applying order of substitution rule and non-substitution rule.

**Comment by vlsi** [comment]
>The main change is here: https://github.com/apache/calcite/pull/2237/files#diff-ad65441646b31d60e7a55d6bac2c1f4d9ddba3473b227963800b1b2bc292a319R342

It must come as its own commit. You should not require each and every reviewer to inspect the diff in order to find a single line change buried under lots of comment changes.

**Comment by chunweilei** [comment]
> It must come as its own commit. You should not require each and every reviewer to inspect the diff in order to find a single line change buried under lots of comment changes.

I do not require anything from anybody. It's just about preference. Does it really so difficult that someone has to spend hours on finding the main change? I don't think so. As you can see, two people have reviewed it at least and give their comments.

**Comment by chunweilei** [comment]
> * change with the corresponding comments

I am open to this change.

**Comment by chunweilei** [comment]
> @chunweilei, please split the change in two:
> 
> * code change with the corresponding comments
> * other comment cleanups
> 
> Currently, it is hard to figure out where is the code part of the change.

I have split the change into two. Thank you for your suggestion.

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 2]
```diff
@@ -335,6 +336,14 @@ default boolean onProduce(RelNode node) {
           physicals.add(rel);
         }
       }
+
+      // No need to apply O_INPUTS if the group has not been implemented yet.
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:340]: What is `O_INPUTS`?

**Comment by vlsi** [comment]
>  have split the change into two

Thanks. It would be better if you specify the nature and the location of the change instead of the current `Some cosmetic cleanup`.
For instance: `Fix grammar in Volcano planner comments`

> It's difficult to add a test to reproduce it. But you can reproduce it if you change the applying order of substitution rule and non-substitution rule. 

If the bug happens only if you manually edit Calcite source, then why do you think it is a bug after all?

>I described it in the jira case.

I do not see where CALCITE-4368 describes why test can't be added.

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:340]: It means OptimzeInput, which is a term in Cascades planner. I borrow it from the comment[1].

[1] https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java#L338

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:340]: I see no reasons to invent obscure words when the thing can be explained in plain English.

**Comment by chunweilei** [comment]
> If the bug happens only if you manually edit Calcite source, then why do you think it is a bug after all?

* It might happen in some cases. But I could not construct one currently.
* Even though we change the order of the rule, the test cases should not fail. It is not expected.

**Comment by chunweilei** [comment]
> I do not see where [CALCITE-4368](https://issues.apache.org/jira/browse/CALCITE-4368) describes why test can't be added.

I mean I describe the way that reproduces the error in the JIRA case.

**Comment by vlsi** [comment]
I'm inclined that this change must be accompanied by a test, otherwise, it should not be committed.

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java (hunk 2 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:340]: I would change it as you suggest. But the term O_INPUT is not invented currently. It comes from a classic paper[1] which was published in 1998.

image


[1] https://15721.courses.cs.cmu.edu/spring2019/papers/22-optimizer1/xu-columbia-thesis1998.pdf

**Comment by chunweilei** [comment]
> I'm inclined that this change must be accompanied by a test, otherwise, it should not be committed.

As you insist, I would try my best to provide one.

**Comment by chunweilei** [comment]
Opened another PR to fix comments: https://github.com/apache/calcite/pull/2243.

**Comment by chunweilei** [comment]
Close the PR since I can not reproduce it without changes.
